ComponentKit/Core/Scope/CKComponentScopeFrame.mm

```diff
@@ -88,4 +88,40 @@ - (instancetype)initWithHandle:(CKComponentScopeHandle *)handle
   return self;
 }
 
+- (void)injectScopeHierarchy:(_CKChildScopeFrameRet)oldFrame
+{
+  Class componentClass = oldFrame.cls;
+  id identifier = oldFrame.identifier;
+  CKAssert([oldFrame.frame isKindOfClass:[CKComponentScopeFrame class]], @"Can't inject with incorrect frame class.");
+
+  auto result = _children.insert({{componentClass, identifier}, oldFrame.frame});
+
+  CKAssert(result.second, @"Scope collision: attempting to create duplicate scope %@:%@", componentClass, identifier);
+}
+
+- (_CKChildScopeFrameRet)childScopeFrameForOldIdentifier:(CKComponentScopeHandleIdentifier)globalIdentifier
+{
+  for (auto &pr : _children) {
+    if (pr.second->_handle.globalIdentifier == globalIdentifier) {
+      return _CKChildScopeFrameRet{pr.first.componentClass, pr.first.identifier, pr.second};
+    }
+  }
+  return {nil, nil, 0};
+}
+
+- (BOOL)anyDescendentHasPendingUpdates:(const CKComponentStateUpdateMap &)stateUpdates
+{
+  auto it = stateUpdates.find(_handle.globalIdentifier);
```

Comment:
Not as far as I can see because `stateUpdates` is just a flat map of ids -> updates.

We could annotate each node in the state tree with a hash-set of all the IDs below it, which would make this an O(1) operation here, but given the typical size of a state tree (10-15 elements) I don't think it's worth avoiding the recursion. For now let's do this, and come back to it if it's a hotspot (doubt it)